From 9c882b2ca0b539cf0e7d127f196a4efa976acb12 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Fri, 21 Jul 2023 14:24:18 +0800
Subject: [PATCH 12/53] [Huawei] mm/sharepool: Delete pvma parameter in
 sp_mmap()

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Update the vma in sp_mmap() and don't return the vma.
This would change the behaviour of sp_group_add_task() but it's
reasonalbe, or the new task for a group could have wirte permission
for the old spa.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 41 ++++++++++++++++++-----------------------
 1 file changed, 18 insertions(+), 23 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 468f2753fedb..f26aa71b58f2 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -849,7 +849,7 @@ static inline bool check_aoscore_process(struct task_struct *tsk)
 
 static unsigned long sp_mmap(struct mm_struct *mm, struct file *file,
 			     struct sp_area *spa, unsigned long *populate,
-			     unsigned long prot, struct vm_area_struct **pvma);
+			     unsigned long prot);
 static void sp_munmap(struct mm_struct *mm, unsigned long addr, unsigned long size);
 
 #define K2U_NORMAL	0
@@ -1495,7 +1495,7 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 			continue;
 		}
 
-		addr = sp_mmap(mm, file, spa, &populate, prot_spa, NULL);
+		addr = sp_mmap(mm, file, spa, &populate, prot_spa);
 		if (IS_ERR_VALUE(addr)) {
 			up_write(&mm->mmap_lock);
 			pr_warn("add group: sp mmap failed %ld\n", (long)addr);
@@ -2076,7 +2076,7 @@ EXPORT_SYMBOL_GPL(mg_sp_free);
 /* wrapper of __do_mmap() and the caller must hold down_write(&mm->mmap_lock). */
 static unsigned long sp_mmap(struct mm_struct *mm, struct file *file,
 			     struct sp_area *spa, unsigned long *populate,
-			     unsigned long prot, struct vm_area_struct **pvma)
+			     unsigned long prot)
 {
 	unsigned long addr = spa->va_start;
 	unsigned long size = spa_size(spa);
@@ -2092,14 +2092,19 @@ static unsigned long sp_mmap(struct mm_struct *mm, struct file *file,
 	if (IS_ERR_VALUE(addr)) {
 		atomic_dec(&spa->use_count);
 		pr_err("do_mmap fails %ld\n", addr);
-	} else {
-		BUG_ON(addr != spa->va_start);
-		vma = find_vma(mm, addr);
-		vma->vm_private_data = spa;
-		if (pvma)
-			*pvma = vma;
+		return addr;
 	}
 
+	BUG_ON(addr != spa->va_start);
+	vma = find_vma(mm, addr);
+	vma->vm_private_data = spa;
+
+	if (prot & PROT_WRITE)
+		/* clean PTE_RDONLY flags or trigger SMMU event */
+		vma->vm_page_prot = __pgprot(((~PTE_RDONLY) & vma->vm_page_prot.pgprot) | PTE_DIRTY);
+	else
+		vma->vm_flags &= ~VM_MAYWRITE;
+
 	return addr;
 }
 
@@ -2215,7 +2220,6 @@ static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
 	/* pass through default permission */
 	unsigned long prot = PROT_READ | PROT_WRITE;
 	unsigned long populate = 0;
-	struct vm_area_struct *vma;
 
 	down_write(&mm->mmap_lock);
 	if (unlikely(mm->core_state)) {
@@ -2231,18 +2235,13 @@ static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
 		prot = PROT_READ;
 
 	/* when success, mmap_addr == spa->va_start */
-	mmap_addr = sp_mmap(mm, spa_file(spa), spa, &populate, prot, &vma);
+	mmap_addr = sp_mmap(mm, spa_file(spa), spa, &populate, prot);
 	if (IS_ERR_VALUE(mmap_addr)) {
 		up_write(&mm->mmap_lock);
 		pr_err("sp mmap in allocation failed %ld\n", mmap_addr);
 		return (int)mmap_addr;
 	}
 
-	if (prot & PROT_WRITE)
-		/* clean PTE_RDONLY flags or trigger SMMU event */
-		vma->vm_page_prot = __pgprot(((~PTE_RDONLY) & vma->vm_page_prot.pgprot) | PTE_DIRTY);
-	else
-		vma->vm_flags &= ~VM_MAYWRITE;
 	up_write(&mm->mmap_lock);
 
 	if (!ac->have_mbind) {
@@ -2476,18 +2475,14 @@ static unsigned long sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *
 	if (kc && (kc->sp_flags & SP_PROT_RO))
 		prot = PROT_READ;
 
-	ret_addr = sp_mmap(mm, spa_file(spa), spa, &populate, prot, &vma);
+	ret_addr = sp_mmap(mm, spa_file(spa), spa, &populate, prot);
 	if (IS_ERR_VALUE(ret_addr)) {
 		pr_debug("k2u mmap failed %lx\n", ret_addr);
 		goto put_mm;
 	}
 
-	if (prot & PROT_WRITE)
-		vma->vm_page_prot = __pgprot(((~PTE_RDONLY) & vma->vm_page_prot.pgprot) | PTE_DIRTY);
-
-	if (kc && (kc->sp_flags & SP_PROT_RO))
-		vma->vm_flags &= ~VM_MAYWRITE;
-
+	/* This should not fail because we hold the mmap_lock sicne mmap */
+	vma = find_vma(mm, spa->va_start);
 	if (is_vm_hugetlb_page(vma)) {
 		ret = remap_vmalloc_hugepage_range(vma, (void *)kva, 0);
 		if (ret) {
-- 
2.17.1

